### ISSUE TITLE
slices.Insert causes out of bounds if there are no values

### ISSUE THREAD

**Issue by sedyh** [golang/go#64152]
### What version of Go are you using (`go version`)?

<pre>
$ go version 
go version go1.21 darwin/arm64
</pre>

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="arm64"
GOBIN=""
GOCACHE="/Users/sedyh/Library/Caches/go-build"
GOENV="/Users/sedyh/Library/Application Support/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="arm64"
GOHOSTOS="darwin"
GOINSECURE=""
GOMODCACHE="/Users/sedyh/Desktop/share/go/path/go1.21.1/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/sedyh/Desktop/share/go/path/go1.21.1"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/Users/sedyh/Desktop/share/go/root/go1.21.1"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/Users/sedyh/Desktop/share/go/root/go1.21.1/pkg/tool/darwin_arm64"
GOVCS=""
GOVERSION="go1.21.1"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/dev/null"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -arch arm64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/x5/kntr8jy137vg9cjlm83chwmh0000gn/T/go-build3952104386=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

### What did you do?

For some reason `slices.Insert` causes out of bounds panic if there no values at `if !overlaps(v, s[i+m:]) {}`.

As far as I can see in https://go.dev/cl/540155, this is intended, but it seems wrong to me, because:
- I don't see any point in this. I think this is the only case why I wanted to use it.
- This case is not processed separately, it is a standard out of bounds panic.
- This is not stated in the function comments.

### What did you expect to see?

```go
s := []float64{}
s = slices.Grow(s, 5)
s = slices.Insert(s, 4, 1.0)
```

The function works the same regardless of how full it is.

### What did you see instead?

panic: runtime error: slice bounds out of range [5:1]

This is not a specially caused panic, it looks like a missed case.

**Comment by randall77** [comment]
What would end up in `s[0]` through `s[3]`?
In other words, what is the slice that you expect to result from the example you gave?

**Comment by sedyh** [comment]
> What would end up in `s[0]` through `s[3]`? In other words, what is the slice that you expect to result from the example you gave?

I hope that slices.Insert would fill them with default values for me (.0) or replace the existing values if index is lower than cap.

**Comment by gophun** [comment]
> * This is not stated in the function comments.

It is stated in the [doc comment](https://pkg.go.dev/slices#Insert): "Insert panics if i is out of range."

> This case is not processed separately, it is a standard out of bounds panic.

Because it is a standard out of bounds panic. `slices.Grow` does not grow the length of a slice, it only grows the capacity. You would get the same panic here:

```
s := []float64{}
s = slices.Grow(s, 5)
s[4] = 1.0
```

**Comment by sedyh** [comment]
I would really like some of the functions in this package to be able to do make/append as needed.

**Comment by sedyh** [comment]
> Is there any workaround or slice trick that combines insert and growing the length?
Expand look like what I need, I wish that func was in the slices package.
```go
a = append(a[:i], append(make([]T, j), a[i:]...)...)
```

**Comment by go101** [comment]
The one-line trick is quite inefficient for some cases.

For OP's specified case:
```
s := []float64{}
s = slices.Grow(s, 5)
s = slices.Insert(s, 4, 1.0)
```

just do:
```
s := append(make([]float64, 4, 5), 1.0)
```
```

**Comment by go101** [comment]
If looks it is essential to put the following line at the start of the `slices.Insert` funciton:

```
_ = s[i:]
```

Otherwise, the panic message is misleading.

**Comment by sedyh** [comment]
> just do:
> ```s := append(make([]float64, 4, 5), 1.0)```

I also need to be able to insert elements at any index location other than extending the slice, so the previous one is more suitable.

**Comment by gazerro** [comment]
I think it could help if the term "out of range" is not used in the documentation:

> ... Insert panics if i is out of range.. ...

because it might not be clear that "out of range" refers to the slicing `s[i:]`. It could be interpreted as referring to the slicing `s[:i]`. With the same value of `i`, `s[i:]` can be out of slice, and `s[:i]` may not be.

To remove this ambiguity, I would write:

> ... Insert panics if i > len(s). ...

**Comment by gophun** [comment]
> because it might not be clear that "out of range" refers to the slicing `s[i:]`. It could be interpreted as referring to the slicing `s[:i]`. With the same value of `i`, `s[i:]` can be out of range, and `s[:i]` may not be.

it simply refers to `s[i]`, not to `s[i:]` or `s[:i]`. Who thinks of slicing when they read "out of range"?

**Comment by callthingsoff** [comment]
> > ... Insert panics if i > len(s). ...

@gazerro Also panics when i < 0.

**Comment by gazerro** [comment]
@callthingsoff You're right. We can use a similar sentence used for `Delete` and `Replace` (see recent commit 7a1fce8751b04f66f758c7aac0efd5937bc670dc), so it would be

> ... Insert panics if i > len(s) or s[:i] is not a valid slice of s. ...

**Comment by callthingsoff** [comment]
> We can use a similar sentence used for `Delete` and `Replace` (see recent commit [7a1fce8](https://github.com/golang/go/commit/7a1fce8751b04f66f758c7aac0efd5937bc670dc)), so it would be
> 
> > ... Insert panics if i > len(s) or s[:i] is not a valid slice of s. ...

cc @ianlancetaylor  @randall77
